File: solutions/nim/allergies/1/allergies.nim
type
  Allergen* = enum
    Eggs, Peanuts, Shellfish, Strawberries, Tomatoes, Chocolate, Pollen, Cats

proc bits(allergen: Allergen): int =
  case allergen:
    of Eggs:
      return 1
    of Peanuts:
      return 2
    of Shellfish:
      return 4
    of Strawberries:
      return 8
    of Tomatoes:
      return 16
    of Chocolate:
      return 32
    of Pollen:
      return 64
    of Cats:
      return 128

proc isAllergicTo*(score: int, allergen: Allergen): bool =
  return (score and bits(allergen)) != 0

proc allergies*(score: int): set[Allergen] =
  var all: set[Allergen] = {}
  for gen in Allergen:
    if isAllergicTo(score, gen):
      all.incl(gen)
  return all

File: solutions/ocaml/allergies/1/allergies.ml
type allergen = Eggs
              | Peanuts
              | Shellfish
              | Strawberries
              | Tomatoes
              | Chocolate
              | Pollen
              | Cats

let allergic_to number allergy =
  let pos = match allergy with
    | Eggs -> 1
    | Peanuts -> 2
    | Shellfish -> 4
    | Strawberries -> 8
    | Tomatoes -> 16
    | Chocolate -> 32
    | Pollen -> 64
    | Cats -> 128
  in
  number land pos > 0

let allergies number =
  let all_allergens = [
    Eggs; Peanuts; Shellfish; Strawberries; Tomatoes; Chocolate; Pollen; Cats
  ] in
  List.filter (fun allergy -> allergic_to number allergy) all_allergens
